### PR TITLE
spread: refactor pipelines and add snapd testing

### PR DESCRIPTION
**Pull request by Meulengracht**
Refactor the test pipeline to seperate build and test stages. 
Include support for running spread tests from snapd (smoke-tests so far)

I will be doing some further fixes, I am initially creating this PR to test the github actions

### CODE REVIEW

File: tests/lib/prepare-utils.sh [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+[Service]
+Type=oneshot
+ExecStart=/usr/lib/snapd/snapd.spread-tests-run-mode-tweaks.sh
+RemainAfterExit=true
+
+[Install]
+WantedBy=multi-user.target
+EOF
+
+cat > "$snapddir/usr/lib/snapd/snapd.spread-tests-run-mode-tweaks.sh" <<'EOF'
```
Comment (valentindavid) [inline tests/lib/prepare-utils.sh:201]: Should not this be indented?

Comment (Meulengracht) [reply tests/lib/prepare-utils.sh:201]: Really good catch, it should be indented!

**Review by alfonsosanchezbeato — APPROVED** [review]
LGTM, thanks

**Review by valentindavid — APPROVED** [review]

File: .github/workflows/tests.yaml [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    runs-on: self-hosted
+    steps:
+      - name: Cleanup job workspace
+        id: cleanup-job-workspace
+        run: |
+          rm -rf "${{ github.workspace }}"
+          mkdir "${{ github.workspace }}"
+      - uses: actions/checkout@v2
+
+      - name: Build snap
```
Comment (valentindavid) [inline .github/workflows/tests.yaml:24]: s/snap/deb/

File: .github/workflows/tests.yaml [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        run: |
+          spread -artifacts=./artifacts google-nested:tests/spread/build/
+          find ./artifacts -type f -name "*.deb" | head -n 1 | while read line ; do cp "$line" "${{ github.workspace }}/uc-initramfs.deb" ; done
+      
+      - uses: actions/upload-artifact@v2
+        with:
+          name: initramfsdeb
+          path: "${{ github.workspace }}/uc-initramfs.deb"
```
Comment (valentindavid) [inline .github/workflows/tests.yaml:32]: It is not really safe in general to build a package in 22, and then install on 20. In this case, the dependencies for the current branch might not be available on ubuntu 20. So a safer thing to do would be to export the pc-kernel snap with initrd inject.

File: .github/workflows/tests.yaml (hunk 2 above)
Comment (valentindavid) [reply .github/workflows/tests.yaml:24]: We should make a snap instead anyway.

**Pull request merged**